File: Labs/Lab8 /Lab8.java
public class Lab8 {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        java.util.Arrays.fill(dp, amount + 1);
        dp[0] = 0;
        for (int i = 1; i <= amount; i++) {
            for (int coin : coins) {
                if (coin <= i) {
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        return dp[amount] > amount ? -1 : dp[amount];
    }

    public static void main(String[] args) {
        Lab8 obj = new Lab8();
        System.out.println(obj.coinChange(new int[] { 1, 2, 5 }, 11)); // Should return 3
        System.out.println(obj.coinChange(new int[] { 2 }, 3)); // Should return -1
        System.out.println(obj.coinChange(new int[] { 1 }, 0)); // Should return 0
    }
}

File: Labs/Lab8 /Lab8.py
from typing import List


class Lab8:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [amount + 1] * (amount + 1)
        dp[0] = 0
        for a in range(1, amount + 1):
            for coin in coins:
                if coin <= a:
                    dp[a] = min(dp[a], dp[a - coin] + 1)
        return dp[amount] if dp[amount] != amount + 1 else -1

# Example usage:
# sol = Solution()
# print(sol.coinChange([1, 2, 5], 11))  # Should return 3
# print(sol.coinChange([2], 3))         # Should return -1
# print(sol.coinChange([1], 0))         # Should return 0
